core/src/main/java/org/elasticsearch/search/slice/SliceBuilder.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.slice;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.action.support.ToXContentToBytes;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.IndexNumericFieldData;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.internal.UidFieldMapper;
+import org.elasticsearch.index.query.QueryParseContext;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ *
+ */
+public class SliceBuilder extends ToXContentToBytes implements Writeable {
+    public static final ParseField FIELD_FIELD = new ParseField("field");
+    public static final ParseField ID_FIELD = new ParseField("id");
+    public static final ParseField MAX_FIELD = new ParseField("max");
+    private final static ObjectParser<SliceBuilder, QueryParseContext> PARSER =
+        new ObjectParser<>("slice", SliceBuilder::new);
+
+    static {
+        PARSER.declareString(SliceBuilder::setField, FIELD_FIELD);
+        PARSER.declareInt(SliceBuilder::setId, ID_FIELD);
+        PARSER.declareInt(SliceBuilder::setMax, MAX_FIELD);
+    }
+
+    /** Name of field to slice against (_uid by default) */
+    private String field = UidFieldMapper.NAME;
+    /** The id of the slice */
+    private int id = -1;
+    /** Max number of slices */
+    private int max = -1;
```

Comment:
IMO this would defeat the purpose. Maybe we could use the circuit breaker or something to fail the query if the memory used by the bitset is too big ?